### PR TITLE
Instance manager live upgrade v2.1

### PR DESCRIPTION
**Pull request by shuo-wu**
https://github.com/longhorn/longhorn/issues/880
https://github.com/longhorn/longhorn/issues/1055

### CODE REVIEW

File: datastore/longhorn.go [hunk 1]
```diff
@@ -1305,29 +1305,67 @@ func CheckInstanceManagerType(im *longhorn.InstanceManager) (types.InstanceManag
...
+	return itemMap, nil
+}
+
+func (s *DataStore) GetInstanceManagerByInstance(obj interface{}) (*longhorn.InstanceManager, error) {
+	var (
+		name, nodeID string
+		imType       types.InstanceManagerType
+	)
+
+	setting, err := s.getSettingRO(string(types.SettingNameDefaultInstanceManagerImage))
```
Comment (yasker) [inline datastore/longhorn.go]: Use `GetSettingValueExisted`

**Comment by yasker** [comment]
You've decided to sync instance manager not in the instance manager controller?

**Comment by shuo-wu** [comment]
Test Case 1:
1. Launch Longhorn system with `shuowu/longhorn-instance-manager:imtest-v2`, which is the same as `longhornio/longhorn-instance-manager:master`.
--> For all instance managers, `Status.MajorVersion` and `Status.MinorVersion` are 1; `Status.Image` is `shuowu/longhorn-instance-manager:imtest-v2`; the labels value are the same as Spec and Status.
2. Launch a pod. Write data into the corresponding Longhorn volume and record the md5sum.
3. Upgrade the system with `shuowu/longhorn-instance-manager:imtest-v3`, which has higher APIVersion(MinorVersion). 
--> All old engine managers except for the one being used by the volume will be removed.
    For all new instance managers, `Status.MajorVersion` and `Status.MinorVersion` are 1 and 2; `Status.Image` is `shuowu/longhorn-instance-manager:imtest-v3`; the labels value are the same as Spec and Status.
    The md5sum of the pod data keep unchanged.
4. Deploy engine image `shuowu/longhorn-engine:imtest-v3`, which is actually the same as `longhornio/longhorn-engine:master`.
5. Apply live upgrade for the volume. --> All old replica managers will be removed.
6. Log in the pod, then check md5sum for the existing data, and write more data. --> The md5sum of the existing data keep unchanged. R/W works fine.
7. Delete and recreate the pod. (Do not delete the PVC) Then check the md5sum. --> All old engine managers will be deleted. The md5sum of the pod data keep unchanged.

Test Case 2:
1. Launch Longhorn system v0.7.0
2. Launch a pod. Write data into the corresponding Longhorn volume and record the md5sum.
3. Upgrade the system with `shuowu/longhorn-instance-manager:imtest-v2`. 
--> All old engine managers except for the one being used by the volume will be removed. 
    For all old existing instance managers, `Status.MajorVersion` and `Status.MinorVersion` are -1; `Status.Image` is `longhornio/longhorn-instance-manager:v0.7.0`; the labels value are the same as Spec and Status.
    For all new instance managers, `Status.MajorVersion` and `Status.MinorVersion` are 1; `Status.Image` is `shuowu/longhorn-instance-manager:imtest-v2`; the labels value are the same as Spec and Status.
4. Log in the pod, check the existing data then write new data. --> The md5sum of the existing data keep unchanged. R/W works fine.
5. Apply live upgrade for the volume. --> Live upgrade won't be triggered. 
6. Delete and recreate the pod. (Do not delete the PVC) Then check the md5sum. --> All old engine managers will be deleted. The md5sum of the pod data keep unchanged.

**Comment by yasker** [comment]
APIMinVersion and APIVersion are not MajorVersion and MinorVersion.

MajorVersion can be 1 when MinorVersion is 0. It's not possible with APIMinVersion and APIVersion.

File: types/resource.go [hunk 2]
```diff
@@ -348,16 +348,21 @@ const (
...
+	// TODO: deprecate this field
+	EngineImage string `json:"engineImage"`
 }
 
 type InstanceManagerStatus struct {
 	OwnerID      string                     `json:"ownerID"`
 	CurrentState InstanceManagerState       `json:"currentState"`
 	Instances    map[string]InstanceProcess `json:"instances"`
 	IP           string                     `json:"ip"`
+	MajorVersion int                        `json:"majorVersion"`
```
Comment (yasker) [inline types/resource.go]: `APIMinVersion` and `APIVersion`

**Pull request merged**